[Serializer] Cache error normalizing anonymous class

**Symfony version(s) affected**: 4.4.7

**Description**  
During phpunit tests with `APP_DEBUG=1`, an anonymous class can be normalized by the `NormalizerInterface`, however, with `APP_DEBUG=0` the following error is produced:

```
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".
```

<details>
<summary>Full stack trace</summary>

```
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".

/app/vendor/symfony/cache/CacheItem.php:177
/app/vendor/symfony/cache/Traits/AbstractTrait.php:281
/app/vendor/symfony/cache/Traits/AbstractAdapterTrait.php:44
/app/vendor/symfony/cache/Adapter/PhpArrayAdapter.php:122
/app/vendor/symfony/serializer/Mapping/Factory/CacheClassMetadataFactory.php:50
/app/vendor/symfony/serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:50
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:269
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:164
/app/vendor/symfony/serializer/Serializer.php:152
/app/tests/AnonymousClassTest.php:19
```

</details>

**How to reproduce**  
<!-- Code and/or config needed to reproduce the problem. If it's a complex bug,
     create a "bug reproducer" as explained in:
     https://symfony.com/doc/current/contributing/code/reproducer.html -->

The error can be reproduced with the following test case:

```php
<?php

namespace App\Tests;

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Symfony\Component\Serializer\Normalizer\NormalizerInterface;

final class AnonymousClassTest extends KernelTestCase
{
    public function testSerializeAnonymousClass(): void
    {
        static::bootKernel();

        $normalizer = static::$container->get(NormalizerInterface::class);

        $class = new class() {
        };

        $normalizer->normalize($class);

        $this->addToAssertionCount(1);
    }
}
```

<details>
<summary>Successful test run with APP_DEBUG=1</summary>

```
$ APP_DEBUG=1 vendor/bin/phpunit tests/AnonymousClassTest.php 
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 466 ms, Memory: 20.00 MB

OK (1 test, 1 assertion)
```

</details>

<details>
<summary>Failed test run with APP_DEBUG=0</summary>

```
$ APP_DEBUG=0 vendor/bin/phpunit tests/AnonymousClassTest.php 
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 216 ms, Memory: 18.00 MB

There was 1 error:

1) App\Tests\AnonymousClassTest::testSerializeAnonymousClass
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".

/app/vendor/symfony/cache/CacheItem.php:177
/app/vendor/symfony/cache/Traits/AbstractTrait.php:281
/app/vendor/symfony/cache/Traits/AbstractAdapterTrait.php:44
/app/vendor/symfony/cache/Adapter/PhpArrayAdapter.php:122
/app/vendor/symfony/serializer/Mapping/Factory/CacheClassMetadataFactory.php:50
/app/vendor/symfony/serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:50
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:269
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:164
/app/vendor/symfony/serializer/Serializer.php:152
/app/tests/AnonymousClassTest.php:19

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

</details>


**Possible Solution**  
<!--- Optional: only if you have suggestions on a fix/reason for the bug -->

It appears that an issue similar to this has appeared a few times in the past (linked issued below) and it was fixed with something like this in `PropertyAccessor::getReadAccessInfo` and `PropertyAccessor::getWriteAccessInfo`:

```
$key = false !== strpbrk($key = $class.'..'.$property, '{}()/@:') ? rawurlencode($key) : $key;
```

But this issue seems to be unrelated to the PropertyAccessor, so I adding the change to `CacheClassMetadataFactory::getMetadataFor` and that makes the test pass successfully with `APP_DEBUG=0`:

```diff
public function getMetadataFor($value)
{
    $class = $this->getClass($value);
    // Key cannot contain backslashes according to PSR-6
    $key = strtr($class, '\\', '_');
+    $key = false !== strpbrk($key, '{}()/@:') ? rawurlencode($key) : $key;

    $item = $this->cacheItemPool->getItem($key);

```

Not sure if this is the correct place to fix the issue, or if it has knock on affects elsewhere.

**Additional context**  
<!-- Optional: any other context about the problem: log messages, screenshots, etc. -->

I found a few older issues which may be related:
* #23136
* #36565 
* #29329 
* #29293 
* #26313

--- a/src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php
+++ b/src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php
@@ -44,8 +44,7 @@ class CacheClassMetadataFactory implements ClassMetadataFactoryInterface
     public function getMetadataFor($value)
     {
         $class = $this->getClass($value);
-        // Key cannot contain backslashes according to PSR-6
-        $key = strtr($class, '\\', '_');
+        $key = rawurlencode(strtr($class, '\\', '_'));
 
         $item = $this->cacheItemPool->getItem($key);
         if ($item->isHit()) {

--- a/src/Symfony/Component/Serializer/Tests/Mapping/Factory/CacheMetadataFactoryTest.php
+++ b/src/Symfony/Component/Serializer/Tests/Mapping/Factory/CacheMetadataFactoryTest.php
@@ -63,4 +63,20 @@ class CacheMetadataFactoryTest extends TestCase
 
         $factory->getMetadataFor('Not\Exist');
     }
+
+    public function testAnonymousClass()
+    {
+        $anonymousObject = new class() {
+        };
+
+        $metadata = new ClassMetadata(\get_class($anonymousObject));
+        $decorated = $this->getMockBuilder(ClassMetadataFactoryInterface::class)->getMock();
+        $decorated
+            ->expects($this->once())
+            ->method('getMetadataFor')
+            ->willReturn($metadata);
+
+        $factory = new CacheClassMetadataFactory($decorated, new ArrayAdapter());
+        $this->assertEquals($metadata, $factory->getMetadataFor($anonymousObject));
+    }
 }

